### PR TITLE
161 add gradient boosting, logistic regression and random forests

### PR DESCRIPTION
**Pull request by nmaarnio**
Basic implementations of gradient boosting, logistic regression and random forests using Sklearn.

### CODE REVIEW

**Comment by nmaarnio** [comment]
@nialov could you take a look at this PR at some point? Also @zelioluca , let me know if you have any comments since you're now also working with Sklearn and modelling stuff. I might be unaware of some best practices in this domain, and I am not sure if the parameter optimization is good like this.

**Comment by nialov** [comment]
Will take a look next week hopefully!

**Comment by zelioluca** [comment]
Perfect man!!!! Thanks

File: eis_toolkit/prediction/gradient_boosting.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+        loss=loss,
+        learning_rate=learning_rate,
+        n_estimators=n_estimators,
+        max_depth=max_depth,
+        subsample=subsample,
+        random_state=random_state,
+        **kwargs,
+    )
+
+    # Training and optionally tuning the model
```
Comment (nialov) [inline eis_toolkit/prediction/gradient_boosting.py:83]: Based on this comment I understand the fitting should be done regardless of using tuning or not. Should update the comment if this is not the case and remove the else-clause if this is the case.

File: eis_toolkit/prediction/gradient_boosting.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+from sklearn.metrics import classification_report
+from sklearn.model_selection import train_test_split
+
+from eis_toolkit import exceptions
+from eis_toolkit.prediction.model_utils import evaluate_regression_model, tune_model_parameters
+
+
+@beartype
+def gradient_boosting_classifier_train(
+    X: Union[np.ndarray, pd.DataFrame],
```
Comment (nialov) [inline eis_toolkit/prediction/gradient_boosting.py:16]: I always find pandas dataframes and series a bit dangerous in comparison to arrays which are usually more simply structured. Probably not worth it at this point to remove functionality but I would have maybe only taken numpy arrays as inputs for the sake of simplicity. 

I also see that dataframe inputs are not tested in the unit tests, I would consider adding a simple test case with a dataframe input instead of an array.

File: eis_toolkit/prediction/model_utils.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+    Returns:
+        Dictionary containing the computed metrics.
+    """
+    mae = mean_absolute_error(y_test, y_pred)
+    mse = mean_squared_error(y_test, y_pred)
+    rmse = np.sqrt(mse)
+    r2 = r2_score(y_test, y_pred)
+
+    metrics = {"MAE": mae, "MSE": mse, "RMSE": rmse, "R2": r2}
```
Comment (nialov) [inline eis_toolkit/prediction/model_utils.py:94]: Nitpicky stuff but I would refactor the strings to module variables e.g. 

~~~python
MAE = "MAE"
MSE = "MSE"
...
~~~

Could then import these in the tests instead of using strings manually there.

File: tests/prediction/gradient_boosting_test.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+from eis_toolkit import exceptions
+from eis_toolkit.prediction.gradient_boosting import (
+    gradient_boosting_classifier_predict,
+    gradient_boosting_classifier_train,
+    gradient_boosting_regressor_predict,
+    gradient_boosting_regressor_train,
+)
+
+X, y = load_iris(return_X_y=True)
```
Comment (nialov) [inline tests/prediction/gradient_boosting_test.py:14]: I would use upper case for these variables to highlight them in the test functions e.g.

~~~python
IRIS_X, IRIS_Y = ...
~~~

**Review by nialov — CHANGES_REQUESTED** [review]
Hey, very small stuff. Fix if you think it is worth the time. I am not an expert on the functionality of these `sklearn` functions so difficult to say anything about the business logic itself but looks very solid to me!

File: eis_toolkit/prediction/gradient_boosting.py (hunk 2 above)
Comment (RichardScottOZ) [reply eis_toolkit/prediction/gradient_boosting.py:16]: On your user side, lots of people don't understand arrays - so having it magically take dataframes/tables and transform them in the background helps some people.

**Comment by zelioluca** [comment]
Hey guys i can help if you need just let me know

File: eis_toolkit/prediction/gradient_boosting.py (hunk 2 above)
Comment (nialov) [reply eis_toolkit/prediction/gradient_boosting.py:16]: If these functions are used as part of a library in interactive Python scripting, you can probably expect the coders to be able to extract arrays from dataframes. You are anyways not passing dataframes or numpy arrays from the QGIS plugin interface as there will anyway be middleware in-between. As I said probably not worth it removing functionality unless some problems actually manifest but simple test cases should be added instead!

**Comment by zelioluca** [comment]
very good job guys! I like it!

**Comment by nmaarnio** [comment]
@msmiyels I have now pushed the reworked versions of these three ML models. The tests should cover most cases, but I might have missed something.

File: tests/prediction/logistic_regression_test.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+from eis_toolkit import exceptions
+from eis_toolkit.prediction.logistic_regression import logistic_regression_train
+
+X_IRIS, Y_IRIS = load_iris(return_X_y=True)
+
+
+def test_logistic_regression():
+    """Test that Logistic Regression works as expected."""
+    metrics = ["accuracy"]
+    model, out_metrics = logistic_regression_train(X_IRIS, Y_IRIS, random_state=42)
```
[the remark this answers is not in the document]
Comment (nmaarnio) [reply tests/prediction/logistic_regression_test.py:16]: So do you think we should choose different parameter values that allows convergence in this case? Perhaps higher value for `max_iter`?

File: tests/prediction/logistic_regression_test.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    predicted_labels = model.predict(X_IRIS)
+
+    assert isinstance(model, LogisticRegression)
+    np.testing.assert_equal(len(predicted_labels), len(Y_IRIS))
+
+    # Test that all predicted labels have perfect metric scores since we are predicting with the test data
```
[the remark this answers is not in the document]
Comment (nmaarnio) [reply tests/prediction/logistic_regression_test.py:22]: Yeah, good points with these tests. However, don't we now already run the model with specified settings, or do you mean you think it's better to write them out, even if they are the default ones? But I agree it's a very good idea to check the outputs and have more precise comparison

File: eis_toolkit/prediction/gradient_boosting.py
[the remark this answers is not in the document]
Comment (nmaarnio) [reply eis_toolkit/prediction/gradient_boosting.py:1]: Does this comment relate to the next one, about random forests and max_depth missing there? I noticed that our docsting said default value for `max_depth` is 100 in gradient boosting, so fixed that anyway :)

File: eis_toolkit/prediction/random_forests.py
[the remark this answers is not in the document]
Comment (nmaarnio) [reply eis_toolkit/prediction/random_forests.py:1]: Good catch! I'll add `max_depth` there with default `None`.

**Comment by nmaarnio** [comment]
Hi Micha,

thanks for the thorough review and good comments!

I'll add both `verbose` mode and `shuffle` to simple train-test split.

About the option to split data before, do you think running a model with `test_method = "none"` does not do the trick here? Of course, if you want to evaluate your model in this case, you need to do it after training the model using a Sklearn method. I might have misunderstood what you meant here, so can you maybe describe it a bit more?

File: tests/prediction/logistic_regression_test.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    for metric in metrics:
+        np.testing.assert_equal(out_metrics[metric], 1.0)
```
[the remark this answers is not in the document]
Comment (nmaarnio) [reply tests/prediction/logistic_regression_test.py:24]: I found out the reason why accuracy is 1.0 in `out_metrics` but 0.98 when it is calculated like in your snippet. The `out_metrics` are produced during training, and in this case a simple split of 20%-80% is used. The model trained with 80% of the data manages to classify correctly all in the 20%, therefore accuracy is 1.0. However, after training the whole data is fit to the model. So both `Y` and `y_pred` in  `score = accuracy_score(Y, y_pred)` are different in your code than what they are in the training function, and model trained with the whole data apparently doesn't manage to predict all its labels correctly.

**Comment by nmaarnio** [comment]
Hey,

no problem.

Okay I understand now. Do you think we should parameterize the training functions so that this whole workflow could be accomplished with one function call? What the user can now do is just as you described, first split their data (using `train_test_split` for example, we can make this importable from EIS or create a wrapper), then train how they like, and finally use `predict`, `score` or some other Sklearn method. Of course, this assumes the user is aware that all these functions/options exist and can be used. 

I guess we could add parameter `evaluate_split` or `validate_split` that would define the other split of data (and maybe swap the names so `validate_split` is the "inner" one and `test_split` for the unseen data), or give the option to give already separated test dataset. However, I am a bit worried these functions will grow too large and complicated to use. In the Plugin, I am sure we can find some ways to show all these different options to the user. But what do you think? I am open to suggestions how to best incorporate this feature :)

**Comment by nmaarnio** [comment]
Hi!

I definitely agree with term consistency – good that you pointed it out! The namings were very unclear now that I revisited them. 

Now:

- core/private function is called `_train_and_validate_sklearn_model` 
- `test_method` is called `validation_method`
- `simple_split` is called just `split` and `simple_split_size` is called `split_size` (I thought the "simple" didn't add much to explain this validation method)
- docstrings are talking only about validation instead of testing / validation

Regarding the workflow for testing/scoring model with unseen data, I think this 3-step workflow is good for the Toolkit side. However, I ended up creating simple EIS Toolkit functions `split_data`, `test_model` and `predict`, so now users can still use our own functions/wrappers to conduct this workflow. What do you think about the structure now?

File: eis_toolkit/prediction/model_utils.py [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+    if validation_method == NO_VALIDATION:
+        model.fit(X, y)
+        metrics = {}
+
+        return model, metrics
+
+    # Approach 2: Validation with splitting data once
+    elif validation_method == SPLIT:
+        X_train, X_valid, y_train, y_valid = split_data(
+            X, y, split_size=split_size, random_state=random_state, shuffle=True
```
[the remark this answers is not in the document]
Comment (nmaarnio) [reply eis_toolkit/prediction/model_utils.py]: I added the shuffle as a parameter to the inner function (`def _train_and_validate_sklearn_model`). I didn't add it yet to the public ones, but can do that too if you think it should be possible.
